src/main/java/com/commercetools/service/paypalPlus/impl/PaypalPlusPaymentServiceImpl.java

```diff
@@ -54,6 +61,47 @@ public PaypalPlusPaymentServiceImpl(@Nonnull APIContextFactory paypalPlusApiCont
         return paymentStageWrapper((paypalPlusApiContext) -> Payment.get(paypalPlusApiContext, paymentId));
     }
 
+    @Override
+    public CompletionStage<Webhook> createWebhook(@Nonnull String notificationUrl) {
+        // https://github.com/paypal/PayPal-Java-SDK/blob/1f85723f3df153e6b876dc2325dcef464af8b6e4/rest-api-sdk/src/test/java/com/paypal/api/payments/WebhookTestCase.java#L105
+        return paymentStageWrapper(EventType::availableEventTypes)
+                .thenCompose(eventTypeList -> paymentStageWrapper(apiContext -> {
+                    Webhook webhook = new Webhook();
+                    webhook.setUrl(notificationUrl);
+                    webhook.setEventTypes(eventTypeList.getEventTypes());
+                    return webhook.create(apiContext, webhook);
+                }));
+    }
+
+    @Override
+    public CompletionStage<Webhook> ensureWebhook(@Nonnull String notificationUrl) {
+        return paymentStageWrapper(apiContext -> {
+            WebhookList webhookList = new WebhookList();
+            webhookList = webhookList.getAll(apiContext);
+            return webhookList.getWebhooks().stream()
+                    .filter(webhook -> notificationUrl.equalsIgnoreCase(webhook.getUrl()))
+                    .findAny();
+        }).thenCompose(webhookOpt -> webhookOpt
+                .map(webhook -> CompletableFuture.completedFuture(webhookOpt.get()))
+                .orElseGet(() -> createWebhook(notificationUrl).toCompletableFuture())
+        );
+    }
+
+    @Override
+    public CompletionStage<Boolean> validateNotificationEvent(@Nonnull Webhook webhook,
+                                                              @Nonnull Map<String, String> headersInfo,
+                                                              @Nonnull String requestBody) {
+        return paymentStageWrapper(apiContext -> {
+            try {
+                apiContext.addConfiguration(Constants.PAYPAL_WEBHOOK_ID, webhook.getId());
+                return Event.validateReceivedEvent(apiContext, headersInfo, requestBody);
+            } catch (InvalidKeyException | NoSuchAlgorithmException | SignatureException e) {
```

Comment:
i'd rather extend `paymentStageWrapper()` catch statement (add new catch entry `catch (InvalidKeyException | NoSuchAlgorithmException | SignatureException e)`) and leave the same `PaypalPlusServiceException` exception, but with some other message, e.g. "Paypal Plus service encryption exception, see the logs"

Reply:
Why such differentiation? For the frontend user, it makes no difference. And for the developer, he/she can check the logs and see the stacktrace there.

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/57/commits/ab4cababbeda738da2731da6069879e5c53ed284

Reply:
Actually, it is opposite - about unification:
instead of having different `try..catch` blocks in different methods, you keep only one (inside `paymentStageWrapper()`, which is desired exactly for this purpose).

Having dedicated catch is not mandatory though, you could catch all of them in the single one. (the only one issue: _InvalidKeyException, NoSuchAlgorithmException, SignatureException are not `PayPalRESTException`, but internal JDK checked exception, that's why i wanted to separate potential errors handling of encrypting out of PaypalREST errors. But not so important though - that are handled the same way, just logged with different message)

Reply:
OK I will leave it as it is for now. There's a task to improve and unify the way we log errors, so it could be done there? https://github.com/commercetools/commercetools-paypal-plus-integration/issues/25

Reply:
i've done it b1e1bda